_posts/2022-07-13-using-the-kani-rust-verifier-on-a-firecracker-example.md

```diff
@@ -0,0 +1,425 @@
+The intention of our mock is to enable Kani to explore the behavior of a caller of `read_obj` (like `parse`) under any of these possibilities.
+
+```rust
+    fn read_obj<T>(&self, addr: GuestAddress) -> Result<T, Error>
+    where
+        T: ByteValued + kani::Invariant + ReadObjChecks<T>,
+    {
+        if kani::any() {
+            let val = kani::any::<T>();
+            T::check_on_read_val(&self, &val);
+            Ok(val)
+        } else {
+            Err(kani::any::<Error>())
+        }
+    }
+```
+
+Now let's address the `ReadObjChecks` trait bound that we added, which allows us to call `T::check_on_read_val` where we pass in the generated symbolic value.
+This allows us to attach our checker to calls of `read_obj` where `T = Descriptor`.
+In this way, we can update our state machine.
+
+```rust
+trait ReadObjChecks<T> {
+    type CheckerType;
+    fn check_on_read_val(mem: &GuestMemoryMmap, read_val: &T);
+}
+
+impl ReadObjChecks<Descriptor> for Descriptor {
+    type CheckerType = DescriptorPermissionChecker;
+    fn check_on_read_val(mem: &GuestMemoryMmap, read_val: &Descriptor) {
+        let current_permission = DescriptorPermission::from_flags(read_val.flags);
+        mem.permission_checker
+            .borrow_mut()
+            .update(current_permission);
+    }
+}
+```
+
+### Proof harness
+
+Putting this altogether, we can write a harness as follows.
+At a high-level, we generate symbolic inputs for the `parse` and then assert our property of interest.
+The only nit is that we must generate an initial valid `DescriptorChain` (not all `kani::any<DescriptorChain>()` values are valid).
+The way we chose to do this is to call `DescriptorChain::checked_new` with symbolic inputs and proceed if this returns a valid `DescriptorChain`.
+
+```rust
+#[cfg(kani)]
+mod verification {
+    use super::*;
+
+    // ANCHOR: stretch_kani
+    #[kani::proof]
+    pub fn requirement_2642() {
+        let mem = GuestMemoryMmap::new();
+        let desc_table: GuestAddress = kani::any();
+        let queue_size: u16 = kani::any();
+        let index: u16 = kani::any();
+        let desc = DescriptorChain::checked_new(&mem, desc_table, queue_size, index);
+        match desc {
+            Some(x) => {
+                let req = Request::parse(&x, &mem, kani::any::<u64>());
+                if req.is_ok() {
+                    assert!(mem.permission_checker.borrow().virtio_2642_holds());
+                }
+                if !(mem.permission_checker.borrow().virtio_2642_holds()) {
+                    assert!(req.is_err());
+                }
+            }
+            None => {}
+        };
+    }
+}
+```
+
+Passing the harness to Kani results in `VERIFICATION:- SUCCESSFUL` in a few seconds.
+As sanity check, if we insert an issue such as forgetting a validity check to ensure the third buffer is read-only then Kani reports an error (since in this case it is possible to fail virtio requirement 2.6.4.2 but return a valid `Request`).
```

Comment:
```suggestion
As sanity check, if we insert an issue such as forgetting a validity check to ensure the third buffer is read-only, then Kani reports an error (since in this case it is possible to fail virtio requirement 2.6.4.2 but return a valid `Request`).
```